### DISCUSSION TITLE
Integrate Headless CMS with Vercel Visual Editing

### DISCUSSION THREAD

**Discussion by liho00** [directus/directus#19677] [Feature Requests]
### Summary

Visual Editing allows you to edit content in real-time on websites using headless Content Management Systems (CMSs). By offering real-time content visualization, it enables collaborators to make changes on preview deployments without requiring a developer's intervention.

Reference:
https://vercel.com/docs/workflow-collaboration/visual-editing

### Basic Example

[Using Visual Editing](https://vercel.com/docs/workflow-collaboration/visual-editing#using-visual-editing)
Once setup, Visual Editing can be enabled on a preview deployment through the toolbar's edit mode button. Using this feature doesn't require any additional configuration or code changes to the page you want to edit.

[Edit mode](https://vercel.com/docs/workflow-collaboration/visual-editing#edit-mode)
When editable information is detected on a page, the toolbar shows an edit button to interact with the editable fields.

Toolbar with edit mode button - Indicated by the pencil icon
Selecting the button activates edit mode, which highlights editable fields in blue as you hover near them. An editable field is determined by it having a corresponding model in the CMS. Hover over an element to display a link in the top-right corner, click to open the related CMS field for editing.

An editable element with a link to open the related field in the CMS

### Motivation

An editable element with a link to open the related field in the CMS
[Streamlining CMS Workflows through Visual Editing](https://vercel.com/docs/workflow-collaboration/visual-editing#streamlining-cms-workflows-through-visual-editing)
Headless CMSs can face challenges in content editing and management, especially in real-time visualization and collaboration. Visual Editing overcomes these challenges by integrating directly into your site and offering the following enhancements:

[Real-time visualization](https://vercel.com/docs/workflow-collaboration/visual-editing#real-time-visualization)
Most CMSs, even with built-in visual editing, don't integrate seamlessly into your site. This lack of integration forces authors to save and refresh the site to see updates, slowing down the editing process. Visual Editing lets authors select a content area on the page and edit it in the CMS. Combined with Incremental Static Regeneration ([ISR](https://vercel.com/docs/incremental-static-regeneration)), authors get near-instant feedback.

[Collaboration](https://vercel.com/docs/workflow-collaboration/visual-editing#collaboration)
Content changes and feedback typically involve back-and-forth communication, causing inefficiencies and potential miscommunication. Visual Editing improves collaboration by allowing team members to make changes and provide feedback on preview deployments, reducing communication overhead.

[Contextual editing](https://vercel.com/docs/workflow-collaboration/visual-editing#contextual-editing)
Content editing in CMSs usually occurs separately from the website's layout and design. This separation makes it hard for authors to visualize their changes. Visual Editing allows authors to edit content within the website's context, offering a clearer understanding of the impact on design and user experience. The ability to jump from content to the editing interface further enhances this experience.

**Comment by rijkvanzanten** [comment]
Yes, please! Lets also see if there's some way we can re-use some of the same concepts to integrate with StackBit (now part of Netlify)

**Reply by cdwmhcc** [reply]
I think it can be used as an extension rather than as a core feature of directus.

**Comment by LasseRosenow** [comment]
I think maybe it would be more convenient to have a `Directus Native Visual Editing` integration that is independent from proprietary Vercel logic and would just work nicely with the Directus Preview?

**Reply by rijkvanzanten** [reply]
It's one of those things where it feels like both not _that_ hard and impossibly difficult at the same time haha

**Comment by BlackDahlia313** [comment]
There is a nuxt extension being made atm for Directus which follows this principal. So I think adding a Vercel/Stackbit one makes sense as there will be alternatives in the marketplace soon as well if you want a self-hosted solution.

**Comment by c-nv-s** [comment]
@rijkvanzanten is there a ticket raised for this feature?

**Reply by rijkvanzanten** [reply]
No, there's not enough community engagement to warrant that unfortunately, see https://docs.directus.io/contributing/feature-request-process.html#request-a-feature
